samples/VirtualSchemes.Basic/Startup.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AuthSamples.VirtualScheme.Basic
+{
+    public class Startup
+    {
+        public Startup(IConfiguration configuration)
+        {
+            Configuration = configuration;
+        }
+
+        public IConfiguration Configuration { get; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddMvc();
+
+            services.AddAuthentication("Request")
+                .AddVirtualScheme("Request", "Request", o =>
+                {
+                    o.DefaultSelector = ctx =>
+                    {
+                        return ctx.Request.Path.StartsWithSegments("/api") ? "Api" : CookieAuthenticationDefaults.AuthenticationScheme;
+                    };
+                })
+                .AddScheme<AuthenticationSchemeOptions, ApiAuthHandler>("Api", o => { })
+                .AddCookie(options =>
+                {
+                    options.AccessDeniedPath = "/account/denied";
+                    options.LoginPath = "/account/login";
+                });
+        }
+
+        public class ApiAuthHandler : AuthenticationHandler<AuthenticationSchemeOptions>
+        {
+            private readonly ClaimsPrincipal _id;
+
+            public ApiAuthHandler(IOptionsMonitor<AuthenticationSchemeOptions> options, ILoggerFactory logger, UrlEncoder encoder, ISystemClock clock) : base(options, logger, encoder, clock)
+            {
+                var id = new ClaimsIdentity("Api");
+                id.AddClaim(new Claim(ClaimTypes.Name, "Hao", ClaimValueTypes.String, "Api"));
+                _id = new ClaimsPrincipal(id);
+            }
+
+            protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+                => Task.FromResult(AuthenticateResult.Success(new AuthenticationTicket(_id, "Api")));
```

Comment:
You could work an API key sample in here. Or better yet write an API Key handler sample in it's own class library and consume it from here.

Reply:
File an issue?  Seems like a good idea but its not directly related to Virtual schemes

Reply:
I'm actually not sure the 'basic' sample should be using cookies either.  I might split things up so the 'basic' sample demonstrates how these work using only fake auth handler schemes.  And then call this sample VirtualSchemes.PathBasedSchemeSelection or something like that

Reply:
https://github.com/aspnet/AuthSamples/issues/8

Reply:
I like it. We're trying to show real usage scenarios and this cookies + API is the most common question we get asked.

Reply:
Okay I guess we can just start with this for now, I'll rename it to PathBasedSchemeSelection and we can deal with any other stuff in future PRs etc.  I'll add this to the list of samples that needs additional functional tests as well (it will basically need to copy the cookie functionals)